fpdf/fpdf.py

```diff
@@ -2251,22 +2254,70 @@ def add_font(
         if family is None:
             family = font_file_path.stem
 
-        fontkey = f"{family.lower()}{style}"
-        # Check if font already added or one of the core fonts
-        if fontkey in self.fonts or fontkey in CORE_FONTS:
-            warnings.warn(
-                f"Core font or font already added '{fontkey}': doing nothing",
-                stacklevel=get_stack_level(),
-            )
-            return
-
         parsed_unicode_range = None
         if unicode_range is not None:
             parsed_unicode_range = get_parsed_unicode_range(unicode_range)
 
-        self.fonts[fontkey] = TTFFont(
-            self, font_file_path, fontkey, style, parsed_unicode_range
-        )
+        def already_exists(fontkey):
+            # Check if font already added or one of the core fonts
+            if fontkey in self.fonts or fontkey in CORE_FONTS:
+                warnings.warn(
+                    f"Core font or font already added '{fontkey}': doing nothing",
+                    stacklevel=get_stack_level(),
+                )
+                return True
+            return False
+
+        style = "".join(sorted(style.upper()))
+        if any(letter not in "BI" for letter in style):
+            raise ValueError(
+                f"Unknown style provided (only B & I letters are allowed): {style}"
+            )
+
+        # Handle variable font.
+        if variations is not None:
+            if not isinstance(variations, dict):
+                raise TypeError("Variations, if specified, must be a dictionary")
+
+            # Check if font is a variable font.
+            fontkey = f"{family.lower()}"
+            var_font = TTFFont(
+                self, font_file_path, fontkey, style, parsed_unicode_range
+            )
+            if "fvar" not in var_font.ttfont:
+                raise AttributeError(f"{fname} is not a variable font")
+
+            # Check variations dictionary
+            if all(
+                key.upper() in ("", "B", "I", "BI") and isinstance(value, dict)
+                for key, value in variations.items()
+            ):
+                for var_style, axes_dict in variations.items():
+                    fontkey = f"{family.lower()}{var_style}"
+                    if already_exists(fontkey):
+                        continue
+                    self.fonts[fontkey] = TTFFont(
+                        self, font_file_path, fontkey, var_style, parsed_unicode_range
+                    )
+
+                    # static=True causes instantiateVariablefont to behave like deprecated fontTools.varLib.mutator.
+                    # and returns a proper static font.
+                    instancer.instantiateVariableFont(
+                        self.fonts[fontkey].ttfont,
+                        axes_dict,
+                        inplace=True,
+                        static=True,
+                    )
```

Comment:
I suggest moving the instancer logic inside `TTFont.__init__()` - all fontTools font handling is already done there.